File: BOJ/2630. 종이의 개수/solution.js
const [n, ...data] = require('fs').readFileSync('/dev/stdin').toString().trim().split('\n');
const mat = data.map(d => d.split(' ').map(Number));
let white = 0, blue = 0;

const check = (x1, y1, x2, y2) => {
    const begin = mat[y1][x1];
    for (let y = y1; y < y2; y += 1) {
        for (let x = x1; x < x2; x += 1) {
            if (begin !== mat[y][x]) return false;
        }
    }
    return true;
}

const solution = (size, x1, y1, x2, y2) => {
    if (check(x1, y1, x2, y2)) {
        mat[y1][x1] ? blue += 1 : white += 1;
        return;
    }

    const half = size >> 1;
    solution(half, x1, y1, x2 - half, y2 - half);
    solution(half, x1 + half, y1, x2, y2 - half);
    solution(half, x1, y1 + half, x2 - half, y2)
    solution(half, x1 + half, y1 + half, x2, y2);
}

solution(+n, 0, 0, +n, +n);
console.log(white);
console.log(blue);

File: BOJ/2630. 종이의 개수/solution.cc
#include <iostream>

using namespace std;

int n, w = 0, b = 0;
int board[128][128];

bool check(int size, int x, int y) {
    int& begin = board[y][x];
    for (int i = y; i < y + size; i += 1) {
        for (int j = x; j < x + size; j += 1) {
            if (board[i][j] != begin) {
                return false;
            }
        }
    }
    return true;
}

void solution(int size, int x, int y) {
    if (check(size, x, y)) {
        board[y][x] ? b += 1 : w += 1;
        return;
    }

    int half = size >> 1;
    solution(half, x, y);
    solution(half, x + half, y);
    solution(half, x, y + half);
    solution(half, x + half, y + half);
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);

    cin >> n;
    for (int y = 0; y < n; y += 1) {
        for (int x = 0; x < n; x += 1) {
            cin >> board[y][x];
        }
    }

    solution(n, 0, 0);

    cout << w << '\n' << b;
    return 0;
}